### ISSUE TITLE
求教！微调后答案不稳定

### ISSUE THREAD

**Issue by wells-Qiang-Chen** [QwenLM/Qwen#1025]
利用少量数据qlora微调qwen-14b-chat-int4，构建rag问询，发现模型输出极不稳定，是否问题会给出不同的答案，想请教是什么问题引起的？使用的是默认的参数，do_sample=False
问题：没有农业银行储蓄卡是向财务部申请开具在职证明吗（检索到的内容一致）
答案1:不是的，如果员工没有中国农业银行储蓄卡，需要在入职后向人力组申请开具在职证明。
答案2:是的，如果员工没有中国农业银行储蓄卡，需要向人力组申请开具在职证明，并经过综合管理部人力组报备财务后，自行前往指定支行（中国农业银行天府路支行）办理。

**Comment by xiaoduozhou** [comment]
设置下：  num_beams=1

**Comment by wells-Qiang-Chen** [comment]
> 设置下： num_beams=1

设置do_sample=False,num_beams=1
还是会有不同的答案：
答案1:是的，如果没有中国农业银行储蓄卡需要在入职后向人力组申请开具在职证明，经综合管理部人力组报备财务后，自行前往指定支行（中国农业银行天府路支行）办理。
答案2:不是的，如果没有中国农业银行储蓄卡需要在入职后向人力组申请开具在职证明，经综合管理部人力组报备财务后，自行前往指定支行（中国农业银行天府路支行）办理。

**Comment by xiaoduozhou** [comment]
> > 设置下： num_beams=1
> 
> 设置do_sample=False,num_beams=1 还是会有不同的答案： 答案1:是的，如果没有中国农业银行储蓄卡需要在入职后向人力组申请开具在职证明，经综合管理部人力组报备财务后，自行前往指定支行（中国农业银行天府路支行）办理。 答案2:不是的，如果没有中国农业银行储蓄卡需要在入职后向人力组申请开具在职证明，经综合管理部人力组报备财务后，自行前往指定支行（中国农业银行天府路支行）办理。

试下把随机数种子设置为固定值，不一定有用
 def set_random_seed(seed):
    import random

    random.seed(seed)

    # pytorch RNGs
    import torch

    torch.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)

    # numpy RNG
    import numpy as np

    np.random.seed(seed)

set_random_seed(666)

**Comment by xiaoduozhou** [comment]
我也遇到了，好像确实都不行，同求解

**Comment by wells-Qiang-Chen** [comment]
> 我也遇到了，好像确实都不行，同求解

好像不微调也会，没搞清楚怎么引起的

**Comment by Elissa0723** [comment]
遇到了相同的问题，对QWEN-14B进行了lora微调，大概测试集中有1%-2%的输出不稳定。如果进行多次预测，会在2个答案中反复横跳。

**Comment by WangJianQ-cmd** [comment]
请问nums_beams在哪个文件中设置？
generations_config.json中好像没有这个参数：
{
  "chat_format": "chatml",
  "do_sample": true,
  "eos_token_id": 151643,
  "max_new_tokens": 16384,
  "max_window_size": 6144,
  "pad_token_id": 151643,
  "repetition_penalty": 1.1,
  "top_k": 0,
  "top_p": 0.8,
  "transformers_version": "4.36.2"
}

**Comment by wells-Qiang-Chen** [comment]
> 请问nums_beams在哪个文件中设置？ generations_config.json中好像没有这个参数： { "chat_format": "chatml", "do_sample": true, "eos_token_id": 151643, "max_new_tokens": 16384, "max_window_size": 6144, "pad_token_id": 151643, "repetition_penalty": 1.1, "top_k": 0, "top_p": 0.8, "transformers_version": "4.36.2" }

在加载模型那里设置
model = AutoModelForCausalLM.from_pretrained(
            path_to_adapter,
            device_map="auto",
            trust_remote_code=True,
            do_sample=False,
            num_beams=1)

**Comment by Elissa0723** [comment]
> > 我也遇到了，好像确实都不行，同求解
> 
> 好像不微调也会，没搞清楚怎么引起的

想问下这个问题解决了吗？

**Comment by wells-Qiang-Chen** [comment]
> > > 我也遇到了，好像确实都不行，同求解
> > 
> > 
> > 好像不微调也会，没搞清楚怎么引起的
> 
> 想问下这个问题解决了吗？

你看一下#1031的回复，好像说是浮点计算会有误差

**Comment by Elissa0723** [comment]
> > > > 我也遇到了，好像确实都不行，同求解
> > > 
> > > 
> > > 好像不微调也会，没搞清楚怎么引起的
> > 
> > 
> > 想问下这个问题解决了吗？
> 
> 你看一下#1031的回复，好像说是浮点计算会有误差

感谢！

**Comment by wells-Qiang-Chen** [comment]
> > > > > 我也遇到了，好像确实都不行，同求解
> > > > 
> > > > 
> > > > 好像不微调也会，没搞清楚怎么引起的
> > > 
> > > 
> > > 想问下这个问题解决了吗？
> > 
> > 
> > 你看一下#1031的回复，好像说是浮点计算会有误差
> 
> 感谢！

搞懂了给我讲一下 哈哈

**Comment by wells-Qiang-Chen** [comment]
> 遇到了相同的问题，对QWEN-14B进行了lora微调，大概测试集中有1%-2%的输出不稳定。如果进行多次预测，会在2个答案中反复横跳。

你用的是量化模型还是非量化模型？有试过不微调会出现吗？我用的量化模型，微调和不微调都会出现

**Comment by Elissa0723** [comment]
> > 遇到了相同的问题，对QWEN-14B进行了lora微调，大概测试集中有1%-2%的输出不稳定。如果进行多次预测，会在2个答案中反复横跳。
> 
> 你用的是量化模型还是非量化模型？有试过不微调会出现吗？我用的量化模型，微调和不微调都会出现

我用的是lora训练的未量化的模型，量化得还没试过

**Comment by Elissa0723** [comment]
> > 遇到了相同的问题，对QWEN-14B进行了lora微调，大概测试集中有1%-2%的输出不稳定。如果进行多次预测，会在2个答案中反复横跳。
> 
> 你用的是量化模型还是非量化模型？有试过不微调会出现吗？我用的量化模型，微调和不微调都会出现

我尝试了量化模型，int8量化后输出变得稳定了。怀疑是训练用的机器和推理用的机器不一致导致的。

**Comment by Wangqi12138** [comment]
@wells-Qiang-Chen  你好，想问一下模型回答不稳定这个有办法解决吗？
